sql: panic due to context cancelation in distsqlrun.(*colBatchScan).Next

**Describe the problem**

A common cause of node deaths when running an overloaded cluster is unsuppressed panics due to context cancellation which occur during query execution. The following stack trace is an example of such a panic:

```
E190502 20:34:16.333597 824414 sql/conn_executor.go:713  [n2,client=10.142.0.123:47354,user=root] a SQL panic has occurred while executing "SELECT count(*) FROM lineitem JOIN supplier ON l_suppkey = s_suppkey": aborted during DistSender.Send: context canceled                                                         
E190502 20:34:16.760274 824414 util/log/crash_reporting.go:204  [n2,client=10.142.0.123:47354,user=root] a panic has occurred!
panic: aborted during DistSender.Send: context canceled [recovered]
        panic: aborted during DistSender.Send: context canceled [recovered]
        panic: panic while executing 1 statements: SELECT count(*) FROM _ JOIN _ ON _ = _; caused by aborted during DistSender.Send: context canceled

goroutine 824414 [running]:
panic(0x31e8b20, 0xc085b3f5f0)
        /usr/local/go/src/runtime/panic.go:556 +0x2cb fp=0xc07dea95b8 sp=0xc07dea9528 pc=0x735dab
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00a9e4880, 0x3afbde0, 0xc032d8d500, 0x2fe20a0, 0xc03b989380)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x36d fp=0xc07dea9678 sp=0xc07dea95b8 pc=0x208c2ad
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00a9e4880, 0x3afbde0, 0xc032d8d500)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61 fp=0xc07dea96b0 sp=0xc07dea9678 pc=0x2209d21
runtime.call32(0x0, 0x34e5310, 0xc00ebb0e90, 0x1800000018)
        /usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc07dea96e0 sp=0xc07dea96b0 pc=0x76497b
panic(0x2fe20a0, 0xc03b989380)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc07dea9770 sp=0xc07dea96e0 pc=0x735c99
github.com/cockroachdb/cockroach/pkg/sql/exec.CatchVectorizedRuntimeError.func1(0xc023497c18)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/error.go:64 +0x4b4 fp=0xc07dea98f8 sp=0xc07dea9770 pc=0x1e3d104
runtime.call32(0x0, 0x34e5ce8, 0xc00d9a1970, 0x800000008)
        /usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc07dea9928 sp=0xc07dea98f8 pc=0x76497b
panic(0x2fe20a0, 0xc03b989380)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc07dea99b8 sp=0xc07dea9928 pc=0x735c99
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*colBatchScan).Next(0xc08b7e6c80, 0x3afbea0, 0xc06f33cd50, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/colbatch_scan.go:55 +0xab fp=0xc07dea9a10 sp=0xc07dea99b8 pc=0x1e7dfcb
github.com/cockroachdb/cockroach/pkg/sql/exec.(*CancelChecker).Next(0xc0627a6f60, 0x3afbea0, 0xc06f33cd50, 0xc076eee020, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/cancel_checker.go:44 +0x69 fp=0xc07dea9a48 sp=0xc07dea9a10 pc=0x1cf8269
github.com/cockroachdb/cockroach/pkg/sql/exec.(*simpleProjectOp).Next(0xc0627a6f80, 0x3afbea0, 0xc06f33cd50, 0xc06f33d2f0, 0x120)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/simple_project.go:81 +0x4b fp=0xc07dea9a80 sp=0xc07dea9a48 pc=0x1e2da6b
github.com/cockroachdb/cockroach/pkg/sql/exec.(*mergeJoinOp).initProberState(0xc02a482900, 0x3afbea0, 0xc06f33cd50)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/mergejoiner.go:533 +0x216 fp=0xc07dea9ac0 sp=0xc07dea9a80 pc=0x1db4b16
github.com/cockroachdb/cockroach/pkg/sql/exec.(*mergeJoinOp).Next(0xc02a482900, 0x3afbea0, 0xc06f33cd50, 0xc08b7e6ec0, 0xc08b7e0400)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/mergejoiner.go:562 +0x85 fp=0xc07dea9b28 sp=0xc07dea9ac0 pc=0x1db4bf5
github.com/cockroachdb/cockroach/pkg/sql/exec.(*simpleProjectOp).Next(0xc0627a71a0, 0x3afbea0, 0xc06f33cd50, 0x3b1f020, 0xc06f33d320)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/simple_project.go:81 +0x4b fp=0xc07dea9b60 sp=0xc07dea9b28 pc=0x1e2da6b
github.com/cockroachdb/cockroach/pkg/sql/exec.(*countOp).Next(0xc06f33d350, 0x3afbea0, 0xc06f33cd50, 0x734ec2, 0xc000000008)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/count.go:61 +0x6b fp=0xc07dea9b98 sp=0xc07dea9b60 pc=0x1cfa12b
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).nextBatch(0xc04d171500)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:109 +0x56 fp=0xc07dea9bd0 sp=0xc07dea9b98 pc=0x1ea4f16
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).nextBatch-fm()
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:116 +0x2a fp=0xc07dea9be8 sp=0xc07dea9bd0 pc=0x1eee3aa
github.com/cockroachdb/cockroach/pkg/sql/exec.CatchVectorizedRuntimeError(0xc028c2bd20, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/error.go:73 +0x55 fp=0xc07dea9c10 sp=0xc07dea9be8 pc=0x1cff9e5
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).Next(0xc04d171500, 0xc08b7e6e00, 0xc000250001, 0xc028c2be88, 0x1ea4e5e)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:116 +0xb4 fp=0xc07dea9e58 sp=0xc07dea9c10 pc=0x1ea5024
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x3afbea0, 0xc06f33cd50, 0x3b065e0, 0xc04d171500, 0x3aeb760, 0xc0082aa000)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:175 +0x35 fp=0xc07dea9e98 sp=0xc07dea9e58 pc=0x1e7cd75
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc04d171500, 0x3afbea0, 0xc06f33cd50)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:808 +0x92 fp=0xc07dea9ed8 sp=0xc07dea9e98 pc=0x1eb00f2
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc03dc71860, 0x3afbea0, 0xc06f33cd50, 0x34e5288, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:626 +0x1e9 fp=0xc07dea9f38 sp=0xc07dea9ed8 pc=0x1e8f979
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc00081e640, 0xc07fa15320, 0xc08f6eaf30, 0xc028c2c898, 0xc0082aa000, 0xc00a9e4c90, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:253 +0x8a0 fp=0xc07deaa788 sp=0xc07dea9f38 pc=0x20f1010
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc00081e640, 0x3afbea0, 0xc06f33c7e0, 0xc00a9e4c90, 0xc07fa15320, 0xc08f6eaf30, 0x3afd760, 0xc03c131400, 0xc0082aa000)                                                                                                                               
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:841 +0x227 fp=0xc07deaaa48 sp=0xc07deaa788 pc=0x20f4767
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00a9e4880, 0x3afbea0, 0xc06f33c7e0, 0xc00a9e4bd0, 0x3, 0x7f14d69d4ac8, 0xc0ea5758c0, 0x1, 0x0, 0x0)                                                                                                                                       
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:923 +0x2d7 fp=0xc07deaaad8 sp=0xc07deaaa48 pc=0x209c857
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc00a9e4880, 0x3afbea0, 0xc06f33c7e0, 0xc00a9e4bd0, 0x7f14d69d4ac8, 0xc0ea5758c0, 0x0, 0x0)                                                                                                                                             
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:736 +0x658 fp=0xc07deaabd8 sp=0xc07deaaad8 pc=0x209b6f8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc00a9e4880, 0x3afbea0, 0xc06f33c7e0, 0x3b016a0, 0xc032d8d700, 0xc00a92c002, 0x44, 0x0, 0xc0002d46c0, 0x1, ...)                                                                                                                               
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:417 +0xb15 fp=0xc07deab3a0 sp=0xc07deaabd8 pc=0x2098725
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc00a9e4880, 0x3afbea0, 0xc06f33c7e0, 0x3b016a0, 0xc032d8d700, 0xc00a92c002, 0x44, 0x0, 0xc0002d46c0, 0x1, ...)                                                                                                                                          
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101 +0x610 fp=0xc07deab648 sp=0xc07deab3a0 pc=0x2097790
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc00a9e4880, 0x3afbea0, 0xc06f33c7b0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1265 +0xf75 fp=0xc07deabe10 sp=0xc07deab648 pc=0x208ebc5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00a9e4880, 0x3afbde0, 0xc032d8d500, 0xc0005349f8, 0x5400, 0x15000, 0xc000534a90, 0xc005245bb0, 0x0, 0x0)                                                                                                                                                    
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1142 +0x1b4 fp=0xc07deabe68 sp=0xc07deabe10 pc=0x208da44
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000c7b500, 0x3afbde0, 0xc032d8d500, 0xc00a9e4880, 0x5400, 0x15000, 0xc000534a90, 0xc005245bb0, 0x0, 0x0)                                                                                                                                                    
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce fp=0xc07deabec8 sp=0xc07deabe68 pc=0x208a79e
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc036b3c7e6, 0xc08b1ea740, 0x3afbde0, 0xc032d8d500, 0xc005245bb0, 0xc000c7b500, 0xc04df9f200, 0x3aff6a0, 0xc08b1ea720, 0xc05172a4e0, ...)                                                                                               
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580 +0x21f fp=0xc07deabf68 sp=0xc07deabec8 pc=0x23007cf
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc07deabf70 sp=0xc07deabf68 pc=0x766691
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:515 +0x17b
```

**To Reproduce**

I produced this specific crashing running a 3-node cluster of n1-standard-8 gce vms running a highly concurrent version of 3 of the following TPC-H queries with data loaded for scale factor 1 and vectorized execution enabled.

Queries:
```
 SELECT count(*) FROM lineitem JOIN supplier ON l_suppkey = s_suppkey
 SELECT count(*) FROM lineitem@l_sk JOIN supplier@primary ON l_suppkey = s_suppkey                                                                                                                                                   
 SELECT count(*) FROM lineitem WHERE l_discount * l_extendedprice > 10000 
```

Workload command:
```
workload run querybench --concurrency 200 --db tpch --query-file queries --ramp 10m --tolerate-errors=t {pgurl:1-3} 
```

**Expected behavior**
No panics or at least handled panics

**Environment:**
 - CockroachDB with a couple unrelated commits off of 42051536d469116ac1fb751c73b9f13b1d24ac9f (specifically https://github.com/ajwerner/cockroach/commits/03283d3f84cc4d1240419115b874e3761315a36a)
 - Server OS: Ubuntu 16.04
 - Client app: `workload`

--- a/pkg/sql/exec/error.go
+++ b/pkg/sql/exec/error.go
@@ -26,14 +26,11 @@ import (
 	"github.com/pkg/errors"
 )
 
-const (
-	panicLineSubstring = "runtime/panic.go"
-	execPackagePrefix  = "github.com/cockroachdb/cockroach/pkg/sql/exec"
-)
+const panicLineSubstring = "runtime/panic.go"
 
 // CatchVectorizedRuntimeError executes operation, catches a runtime error if
-// it is coming from exec package, and returns it. If an error occurs that is
-// not from exec package, it is not recovered from.
+// it is coming from the vectorized engine, and returns it. If an error not
+// related to the vectorized engine occurs, it is not recovered from.
 func CatchVectorizedRuntimeError(operation func()) (retErr error) {
 	defer func() {
 		if err := recover(); err != nil {
@@ -50,8 +47,10 @@ func CatchVectorizedRuntimeError(operation func()) (retErr error) {
 				panic(fmt.Sprintf("panic line %q not found in the stack trace\n%s", panicLineSubstring, stackTrace))
 			}
 			if scanner.Scan() {
-				if strings.HasPrefix(strings.TrimSpace(scanner.Text()), execPackagePrefix) {
-					// We only want to catch runtime errors coming from the exec package.
+				panicEmittedFrom := strings.TrimSpace(scanner.Text())
+				if isPanicFromVectorizedEngine(panicEmittedFrom) {
+					// We only want to catch runtime errors coming from the vectorized
+					// engine.
 					switch t := err.(type) {
 					case *pgerror.Error:
 						retErr = t
@@ -72,6 +71,20 @@ func CatchVectorizedRuntimeError(operation func()) (retErr error) {
 	}()
 	operation()
 	return retErr
+}
+
+const (
+	execPackagePrefix  = "github.com/cockroachdb/cockroach/pkg/sql/exec"
+	colBatchScanPrefix = "github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*colBatchScan)"
+)
+
+// isPanicFromVectorizedEngine checks whether the panic that was emitted from
+// panicEmittedFrom line of code (which includes package name as well as the
+// file name and the line number) came from the vectorized engine.
+// panicEmittedFrom must be trimmed to not have any white spaces in the prefix.
+func isPanicFromVectorizedEngine(panicEmittedFrom string) bool {
+	return strings.HasPrefix(panicEmittedFrom, execPackagePrefix) ||
+		strings.HasPrefix(panicEmittedFrom, colBatchScanPrefix)
 }
 
 // TestVectorizedErrorEmitter is an Operator that panics on every odd-numbered

